// amazeui-react/lib/Alert.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');
var omit = require('object.omit');
var ClassNameMixin = require('./mixins/ClassNameMixin');

var Alert = React.createClass({
  displayName: 'Alert',

  mixins: [ClassNameMixin],

  propTypes: {
    classPrefix: React.PropTypes.string.isRequired,
    amStyle: React.PropTypes.oneOf(['secondary', 'success', 'warning', 'danger']),
    onClose: React.PropTypes.func
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'alert'
    };
  },

  renderCloseBtn: function renderCloseBtn() {
    return React.createElement(
      'button',
      {
        type: 'button',
        className: this.setClassNamespace('close'),
        onClick: this.props.onClose
      },
      '\xD7'
    );
  },

  render: function render() {
    var classSet = this.getClassSet();
    var restProps = omit(this.props, Object.keys(this.constructor.propTypes));
    var isCloseable = !!this.props.onClose;

    if (this.props.amStyle) {
      classSet[this.prefixClass(this.props.amStyle)] = true;
    }

    classSet[this.prefixClass('closeable')] = isCloseable;

    return React.createElement(
      'div',
      _extends({}, restProps, {
        className: classNames(this.props.className, classSet)
      }),
      isCloseable ? this.renderCloseBtn() : null,
      this.props.children
    );
  }
});

module.exports = Alert;
