// amazeui-react/lib/Divider.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');
var omit = require('object.omit');
var ClassNameMixin = require('./mixins/ClassNameMixin');

var Divider = React.createClass({
  displayName: 'Divider',

  mixins: [ClassNameMixin],

  propTypes: {
    theme: React.PropTypes.oneOf(['default', 'dotted', 'dashed']),
    classPrefix: React.PropTypes.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'divider',
      theme: 'default'
    };
  },

  render: function render() {
    var classSet = this.getClassSet();
    var restProps = omit(this.props, Object.keys(this.constructor.propTypes));

    return React.createElement('hr', _extends({}, restProps, {
      'data-am-widget': this.props.classPrefix,
      className: classNames(this.props.className, classSet)
    }));
  }
});

module.exports = Divider;
