// amazeui-react/lib/Dropdown.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var ReactDOM = require('react-dom');
var classNames = require('classnames');
var assign = require('object-assign');
var omit = require('object.omit');
var ClassNameMixin = require('./mixins/ClassNameMixin');
var constants = require('./constants');
var Button = require('./Button');
var Icon = require('./Icon');
var Events = require('./utils/Events');
var isNodeInTree = require('./utils/isNodeInTree');
var createChainedFunction = require('./utils/createChainedFunction');
var canUseDOM = require('./utils/canUseDOM');

var Dropdown = React.createClass({
  displayName: 'Dropdown',

  mixins: [ClassNameMixin],

  propTypes: {
    title: React.PropTypes.node.isRequired,
    dropup: React.PropTypes.bool,
    navItem: React.PropTypes.bool,
    btnStyle: React.PropTypes.string,
    btnSize: React.PropTypes.string,
    btnInlineStyle: React.PropTypes.object,
    contentInlineStyle: React.PropTypes.object,
    contentComponent: React.PropTypes.node,
    toggleClassName: React.PropTypes.string,
    caretClassName: React.PropTypes.string,
    contentClassName: React.PropTypes.string,
    onOpen: React.PropTypes.func, // open callback
    onClose: React.PropTypes.func // close callback
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'dropdown',
      contentComponent: 'ul'
    };
  },

  getInitialState: function getInitialState() {
    return {
      open: false
    };
  },

  componentWillMount: function componentWillMount() {
    this.unbindOuterHandlers();
  },

  componentWillUnmount: function componentWillUnmount() {
    this.unbindOuterHandlers();
  },

  /**
   * setDropdown
   * @param {bool} isOpen - Dropdown state, `true` -> open, `close` -> false
   * @param {function} [callback]
   */
  setDropdown: function setDropdown(isOpen, callback) {
    if (isOpen) {
      this.bindOuterHandlers();
    } else {
      this.unbindOuterHandlers();
    }

    this.setState({
      open: isOpen
    }, function () {
      callback && callback();

      isOpen && this.props.onOpen && this.props.onOpen();
      !isOpen && this.props.onClose && this.props.onClose();
    });
  },

  // close dropdown on `esc` keyup
  handleKeyup: function handleKeyup(e) {
    e.keyCode === 27 && this.setDropdown(false);
  },

  // close dropdown when click outer dropdown
  handleOuterClick: function handleOuterClick(e) {
    if (isNodeInTree(e.target, ReactDOM.findDOMNode(this))) {
      return false;
    }

    this.setDropdown(false);
  },

  bindOuterHandlers: function bindOuterHandlers() {
    if (canUseDOM) {
      Events.on(document, 'click', this.handleOuterClick);
      Events.on(document, 'keyup', this.handleKeyup);
    }
  },

  unbindOuterHandlers: function unbindOuterHandlers() {
    if (canUseDOM) {
      Events.off(document, 'click', this.handleOuterClick);
      Events.off(document, 'keyup', this.handleKeyup);
    }
  },

  handleDropdownClick: function handleDropdownClick(e) {
    e.preventDefault();

    this.setDropdown(!this.state.open);
  },

  renderChildren: function renderChildren() {
    var _this = this;

    return React.Children.map(this.props.children, function (child, index) {
      if (React.isValidElement(child)) {
        var closeOnClick = child.props.closeOnClick;
        var onClick = child.props.onClick;
        var handleClick = closeOnClick ? createChainedFunction(onClick, function () {
          _this.setDropdown(false);
        }) : onClick;

        return React.cloneElement(child, assign({}, child.props, {
          key: 'dropdownItem-' + index,
          onClick: handleClick
        }));
      }
    });
  },

  render: function render() {
    var classSet = this.getClassSet();
    var Component = this.props.navItem ? 'li' : 'div';
    var btnClassPrefix = this.props.navItem ? '' : 'btn';
    var btnComponent = this.props.navItem ? 'a' : null;
    var caret = React.createElement(Icon, {
      className: this.props.caretClassName,
      icon: 'caret-' + (this.props.dropup ? 'up' : 'down')
    });
    var animation = this.state.open ? this.setClassNamespace('animation-slide-top-fixed') : this.setClassNamespace('dropdown-animation');
    var ContentComponent = this.props.contentComponent;

    classSet[constants.CLASSES.active] = this.state.open;
    classSet[this.prefixClass('up')] = this.props.dropup;

    return React.createElement(
      Component,
      {
        className: classNames(this.props.className, classSet)
      },
      React.createElement(
        Button,
        {
          onClick: this.handleDropdownClick,
          amStyle: this.props.btnStyle,
          amSize: this.props.btnSize,
          style: this.props.btnInlineStyle,
          className: classNames(this.prefixClass('toggle'), this.props.toggleClassName),
          classPrefix: btnClassPrefix,
          component: btnComponent,
          ref: 'dropdownToggle'
        },
        this.props.title,
        ' ',
        caret
      ),
      React.createElement(
        ContentComponent,
        {
          ref: 'dropdownContent',
          style: this.props.contentInlineStyle,
          className: classNames(this.prefixClass('content'), animation, this.props.contentClassName)
        },
        this.renderChildren()
      )
    );
  }
});

Dropdown.Item = React.createClass({
  displayName: 'Item',

  mixins: [ClassNameMixin],

  propTypes: {
    closeOnClick: React.PropTypes.bool,
    href: React.PropTypes.string,
    target: React.PropTypes.string,
    title: React.PropTypes.string,
    header: React.PropTypes.bool,
    divider: React.PropTypes.bool,
    linkComponent: React.PropTypes.any,
    linkProps: React.PropTypes.object,
    active: React.PropTypes.bool
  },

  render: function render() {
    var classSet = this.getClassSet();
    var restProps = omit(this.props, Object.keys(this.constructor.propTypes));
    var children = null;

    classSet[this.setClassNamespace('dropdown-header')] = this.props.header;

    if (this.props.header) {
      children = this.props.children;
    } else if (!this.props.divider) {
      var Component = this.props.linkComponent || 'a';

      children = React.createElement(
        Component,
        _extends({
          onClick: this.handleClick,
          href: this.props.href,
          target: this.props.target,
          title: this.props.title
        }, this.props.linkProps),
        this.props.children
      );
    }

    return React.createElement(
      'li',
      _extends({}, restProps, {
        title: null,
        href: null,
        className: classNames(this.props.className, classSet)
      }),
      children
    );
  }
});

module.exports = Dropdown;

/*
* TODO:
*   1. 关闭动画
*   2. 位置检测/宽度适应
* */
