// amazeui-react/lib/Footer.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');
var omit = require('object.omit');
var ClassNameMixin = require('./mixins/ClassNameMixin');

var Footer = React.createClass({
  displayName: 'Footer',

  mixins: [ClassNameMixin],

  propTypes: {
    theme: React.PropTypes.oneOf(['default']),
    classPrefix: React.PropTypes.string,
    mobileTitle: React.PropTypes.string,
    mobileLink: React.PropTypes.string,
    desktopTitle: React.PropTypes.string,
    desktopLink: React.PropTypes.string,
    onRequestMobile: React.PropTypes.func,
    onRequestDesktop: React.PropTypes.func,
    data: React.PropTypes.array
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'footer',
      theme: 'default',
      mobileTitle: '适配版',
      desktopTitle: '电脑版'
    };
  },

  render: function render() {
    var classSet = this.getClassSet();
    var MobileTag = this.props.mobileLink ? 'a' : 'span';
    var restProps = omit(this.props, Object.keys(this.constructor.propTypes));

    return React.createElement(
      'footer',
      _extends({}, restProps, {
        'data-am-widget': this.props.classPrefix,
        className: classNames(this.props.className, classSet)
      }),
      React.createElement(
        'div',
        { className: this.prefixClass('switch') },
        React.createElement(
          MobileTag,
          {
            className: this.prefixClass('ysp'),
            onClick: this.props.onRequestMobile,
            href: this.props.mobileLink,
            'data-rel': 'mobile'
          },
          this.props.mobileTitle
        ),
        React.createElement(
          'span',
          { className: this.prefixClass('divider') },
          '|'
        ),
        React.createElement(
          'a',
          {
            'data-rel': 'desktop',
            href: this.props.desktopLink,
            onClick: this.props.onRequestDesktop,
            className: this.prefixClass('desktop')
          },
          this.props.desktopTitle
        )
      ),
      React.createElement(
        'div',
        { className: this.prefixClass('miscs') },
        this.props.data ? this.props.data.map(function (item, i) {
          return React.createElement(
            'p',
            { key: i },
            item
          );
        }) : this.props.children
      )
    );
  }
});

module.exports = Footer;
