// amazeui-react/lib/Form.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');
var omit = require('object.omit');
var ClassNameMixin = require('./mixins/ClassNameMixin');

var Form = React.createClass({
  displayName: 'Form',

  mixins: [ClassNameMixin],

  propTypes: {
    classPrefix: React.PropTypes.string.isRequired,
    horizontal: React.PropTypes.bool,
    inline: React.PropTypes.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      classPrefix: 'form'
    };
  },

  render: function render() {
    var inline = this.props.inline;
    var restProps = omit(this.props, Object.keys(this.constructor.propTypes));
    var classSet = {};

    // remove .am-form className if `inline` prop set
    if (inline) {
      classSet[this.prefixClass('inline')] = true;
    } else {
      classSet = this.getClassSet();
      classSet[this.prefixClass('horizontal')] = this.props.horizontal;
    }

    return React.createElement(
      'form',
      _extends({}, restProps, {
        className: classNames(classSet, this.props.className)
      }),
      this.props.children
    );
  }
});

module.exports = Form;
