// @icedesign/container/lib/IceContainer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _loading = _interopRequireDefault(require("@icedesign/base/lib/loading"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Container =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Container, _Component);

  function Container() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Container);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Container)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _this.renderChildren = function () {
      var _this$props = _this.props,
          error = _this$props.error,
          empty = _this$props.empty,
          children = _this$props.children;

      if (error) {
        return _react.default.createElement("div", {
          style: {
            padding: '80px 0',
            textAlign: 'center'
          }
        }, _react.default.createElement("img", {
          style: {
            width: '108px'
          },
          src: "https://img.alicdn.com/tfs/TB1KJkbRFXXXXbRXVXXXXXXXXXX-216-218.png",
          alt: "\u6570\u636E\u52A0\u8F7D\u9519\u8BEF"
        }), _react.default.createElement("p", {
          style: {
            width: '80%',
            margin: '30px auto 0',
            color: '#999',
            textAlign: 'center'
          }
        }, error));
      } else if (empty) {
        return _react.default.createElement("div", {
          style: {
            padding: '80px 0',
            textAlign: 'center'
          }
        }, _react.default.createElement("img", {
          style: {
            width: '97px'
          },
          src: "https://img.alicdn.com/tfs/TB1df3oRFXXXXbEXFXXXXXXXXXX-194-220.png",
          alt: "\u6570\u636E\u4E3A\u7A7A"
        }), _react.default.createElement("p", {
          style: {
            width: '80%',
            margin: '30px auto 0',
            color: '#999',
            textAlign: 'center'
          }
        }, empty));
      }

      return children;
    };

    return _this;
  }

  (0, _createClass2.default)(Container, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          loading = _this$props2.loading,
          children = _this$props2.children,
          title = _this$props2.title,
          style = _this$props2.style,
          className = _this$props2.className,
          error = _this$props2.error,
          empty = _this$props2.empty,
          others = (0, _objectWithoutProperties2.default)(_this$props2, ["loading", "children", "title", "style", "className", "error", "empty"]);
      var containerStyle = (0, _objectSpread2.default)({
        backgroundColor: '#fff',
        borderRadius: '6px',
        padding: '20px',
        marginBottom: '20px'
      }, style);

      if (loading) {
        return _react.default.createElement(_loading.default, {
          shape: "fusion-reactor",
          color: "#66AAFF",
          style: {
            width: '100%'
          }
        }, _react.default.createElement("div", {
          className: "container-block ".concat(className),
          style: containerStyle
        }, children));
      }

      return _react.default.createElement("div", (0, _extends2.default)({
        className: "container-block ".concat(className),
        style: containerStyle
      }, others), title && _react.default.createElement("h4", {
        style: {
          height: '16px',
          lineHeight: '16px',
          fontSize: '16px',
          color: '#333',
          fontWeight: 'bold',
          margin: 0,
          padding: 0,
          marginBottom: '20px'
        }
      }, title), this.renderChildren());
    }
  }]);
  return Container;
}(_react.Component);

exports.default = Container;
Container.displayName = 'Container';
Container.propTypes = {
  /**
   * 加载的loading
   */
  loading: _propTypes.default.bool,

  /**
   * 数据错误
   */
  error: _propTypes.default.any,

  /**
   * 数据为空
   */
  empty: _propTypes.default.any,

  /**
   * 样式
   */
  style: _propTypes.default.object,

  /**
   * 样式名
   */
  className: _propTypes.default.string,

  /**
   * 标题
   */
  title: _propTypes.default.string
};
Container.defaultProps = {
  loading: false,
  error: false,
  empty: false,
  style: {},
  className: '',
  title: ''
};
