// @icedesign/layout/lib/AsideTrigger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _icon = _interopRequireDefault(require("@icedesign/base/lib/icon"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var IceLayoutAsideTrigger =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(IceLayoutAsideTrigger, _Component);

  function IceLayoutAsideTrigger() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, IceLayoutAsideTrigger);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(IceLayoutAsideTrigger)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _this.toggleCollapse = function () {
      var newCollapsed = !_this.props.collapsed;

      _this.props.onClick(newCollapsed);
    };

    return _this;
  }

  (0, _createClass2.default)(IceLayoutAsideTrigger, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          collapsed = _this$props.collapsed,
          style = _this$props.style;
      var classes = (0, _classnames.default)('ice-layout-aside-trigger', {
        'ice-layout-aside-trigger-collapsed': collapsed
      }, className);
      return _react.default.createElement("div", {
        onClick: this.toggleCollapse,
        className: classes,
        style: style
      }, _react.default.createElement(_icon.default, {
        type: "arrow-left"
      }));
    }
  }]);
  return IceLayoutAsideTrigger;
}(_react.Component);

exports.default = IceLayoutAsideTrigger;
IceLayoutAsideTrigger.displayName = 'IceLayoutAsideTrigger';
IceLayoutAsideTrigger.propTypes = {
  collapsed: _propTypes.default.bool
};
IceLayoutAsideTrigger.defaultProps = {
  collapsed: false
};
