// @icedesign/layout/lib/Header.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var IceLayoutHeader =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(IceLayoutHeader, _Component);

  function IceLayoutHeader() {
    (0, _classCallCheck2.default)(this, IceLayoutHeader);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(IceLayoutHeader).apply(this, arguments));
  }

  (0, _createClass2.default)(IceLayoutHeader, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          alignItems = _this$props.alignItems,
          style = _this$props.style,
          className = _this$props.className,
          children = _this$props.children,
          theme = _this$props.theme,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["alignItems", "style", "className", "children", "theme"]);
      var classes = (0, _classnames.default)('ice-layout-header', (0, _defineProperty2.default)({}, "ice-layout-theme-".concat(theme), theme), className);
      var divStyle = (0, _objectSpread2.default)({}, style, {
        alignItems: alignItems
      });
      return _react.default.createElement("div", (0, _extends2.default)({}, others, {
        className: classes,
        style: divStyle
      }), children);
    }
  }]);
  return IceLayoutHeader;
}(_react.Component);

exports.default = IceLayoutHeader;
IceLayoutHeader.displayName = 'IceLayoutHeader';
IceLayoutHeader.propTypes = {
  /**
   * 主题
   */
  theme: _propTypes.default.oneOf(['dark', 'light']),

  /**
   * 子元素垂直对齐方式
   */
  alignItems: _propTypes.default.oneOf(['flex-start', 'flex-end', 'center', 'baseline', 'stretch'])
};
IceLayoutHeader.defaultProps = {
  theme: 'light',
  alignItems: 'center'
};
