// @icedesign/balloon-confirm/lib/BalloonConfirm.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("@alifd/next/es/icon/style");

var _icon = _interopRequireDefault(require("@alifd/next/es/icon"));

require("@alifd/next/es/balloon/style");

var _balloon = _interopRequireDefault(require("@alifd/next/es/balloon"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

require("@alifd/next/es/button/style");

var _button = _interopRequireDefault(require("@alifd/next/es/button"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var BalloonConfirm =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(BalloonConfirm, _Component);

  function BalloonConfirm(props) {
    var _this;

    (0, _classCallCheck2.default)(this, BalloonConfirm);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(BalloonConfirm).call(this, props));

    _this.handleCanel = function (e) {
      var onCancel = _this.props.onCancel;
      onCancel(e);

      _this.closeBallon();
    };

    _this.handleOk = function (e) {
      var onConfirm = _this.props.onConfirm;
      onConfirm(e);

      _this.closeBallon();
    };

    _this.closeBallon = function () {
      _this.confirmRef.current.getInstance().setState({
        visible: false
      });
    };

    _this.confirmRef = _react.default.createRef();
    return _this;
  }

  (0, _createClass2.default)(BalloonConfirm, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          className = _this$props.className,
          title = _this$props.title,
          confirmText = _this$props.confirmText,
          cancelText = _this$props.cancelText,
          IconComponent = _this$props.Icon,
          onConfirm = _this$props.onConfirm,
          onCancel = _this$props.onCancel,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["children", "className", "title", "confirmText", "cancelText", "Icon", "onConfirm", "onCancel"]);
      return _react.default.createElement(_balloon.default, (0, _extends2.default)({
        triggerType: "click",
        closable: false
      }, others, {
        className: "ice-ballon-confirm".concat(className ? " ".concat(className) : ''),
        trigger: children,
        ref: this.confirmRef
      }), _react.default.createElement("div", {
        className: "ice-ballon-confirm-title"
      }, IconComponent, title), _react.default.createElement("div", {
        className: "ice-ballon-confirm-btn-group"
      }, _react.default.createElement(_button.default, {
        size: "small",
        onClick: this.handleCanel
      }, cancelText), _react.default.createElement(_button.default, {
        size: "small",
        type: "primary",
        onClick: this.handleOk
      }, confirmText)));
    }
  }]);
  return BalloonConfirm;
}(_react.Component);

exports.default = BalloonConfirm;
BalloonConfirm.displayName = 'BalloonConfirm';
BalloonConfirm.propTypes = {
  /**
   * 样式名
   */
  className: _propTypes.default.string,

  /**
   * 确认按钮文字
   */
  confirmText: _propTypes.default.string,

  /**
   * 取消按钮文字
   */
  cancelText: _propTypes.default.string,

  /**
   * 确认框描述
   */
  title: _propTypes.default.string,

  /**
   * 点击取消的回调
   */
  onCancel: _propTypes.default.func,

  /**
   * 确认框描述
   */
  onConfirm: _propTypes.default.func,

  /**
   * 自定义Icon
   */
  Icon: _propTypes.default.element
};
BalloonConfirm.defaultProps = {
  className: undefined,
  confirmText: '确定',
  cancelText: '取消',
  title: '',
  onCancel: function onCancel() {},
  onConfirm: function onConfirm() {},
  Icon: _react.default.createElement(_icon.default, {
    type: "warning",
    style: {
      color: '#FFA003'
    }
  })
};
