// @icedesign/img/lib/modules/Contain.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _aliCDNSuffix = require('./aliCDNSuffix');

var _aliCDNSuffix2 = _interopRequireDefault(_aliCDNSuffix);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Contain = (_temp2 = _class = function (_Component) {
  _inherits(Contain, _Component);

  function Contain() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Contain);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Contain.__proto__ || Object.getPrototypeOf(Contain)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Contain, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          alt = _props.alt,
          title = _props.title,
          src = _props.src,
          width = _props.width,
          height = _props.height,
          _props$style = _props.style,
          style = _props$style === undefined ? {} : _props$style,
          className = _props.className,
          onError = _props.onError,
          shape = _props.shape,
          enableAliCDNSuffix = _props.enableAliCDNSuffix;


      var styles = _extends({}, style, {
        display: 'inline-block',
        width: width,
        height: height,
        overflow: 'hidden'
      });

      var centerStyles = {
        width: width,
        height: height,
        display: 'table-cell',
        verticalAlign: 'middle',
        textAlign: 'center'
      };

      var imgStyles = {
        width: 'auto',
        height: 'auto',
        maxWidth: width,
        maxHeight: height
      };

      var realSrc = enableAliCDNSuffix ? src + (0, _aliCDNSuffix2.default)({ width: width, height: height }) : src;

      var cls = (0, _classnames2.default)('ice-img', shape, className);

      return _react2.default.createElement(
        'div',
        { className: cls, style: styles },
        _react2.default.createElement(
          'div',
          { style: centerStyles },
          _react2.default.createElement('img', {
            style: imgStyles,
            onError: onError,
            ref: 'img',
            src: realSrc,
            alt: alt,
            title: title
          })
        )
      );
    }
  }]);

  return Contain;
}(_react.Component), _class.displayName = 'Contain', _class.propTypes = {
  src: _propTypes2.default.string.isRequired,
  width: _propTypes2.default.number,
  height: _propTypes2.default.number,
  alt: _propTypes2.default.string,
  onError: _propTypes2.default.func,
  style: _propTypes2.default.object,
  shape: _propTypes2.default.string
}, _class.defaultProps = {
  onError: function onError() {},
  alt: ''
}, _temp2);
exports.default = Contain;
