// @icedesign/base/index.js
exports.Accordion = require('./lib/_components/@alife/next-accordion/lib/index.js');
exports.Affix = require('./lib/_components/@alife/next-affix/lib/index.js');
exports.Animate = require('./lib/_components/@alife/next-animate/lib/index.js');
exports.Badge = require('./lib/_components/@alife/next-badge/lib/index.js');
exports.Balloon = require('./lib/_components/@alife/next-balloon/lib/index.js');
exports.Breadcrumb = require('./lib/_components/@alife/next-breadcrumb/lib/index.js');
exports.Button = require('./lib/_components/@alife/next-button/lib/index.js');
exports.Calendar = require('./lib/_components/@alife/next-calendar/lib/index.js');
exports.Card = require('./lib/_components/@alife/next-card/lib/index.js');
exports.Cascader = require('./lib/_components/@alife/next-cascader/lib/index.js');
exports.CascaderSelect = require('./lib/_components/@alife/next-cascader-select/lib/index.js');
exports.Checkbox = require('./lib/_components/@alife/next-checkbox/lib/index.js');
exports.ConfigProvider = require('./lib/_components/@alife/next-config-provider/lib/index.js');
exports.DatePicker = require('./lib/_components/@alife/next-date-picker/lib/index.js');
exports.Dialog = require('./lib/_components/@alife/next-dialog/lib/index.js');
exports.dom = require('./lib/_components/@alife/next-dom/lib/index.js');
exports.Dropdown = require('./lib/_components/@alife/next-dropdown/lib/index.js');
exports.Feedback = require('./lib/_components/@alife/next-feedback/lib/index.js');
exports.Field = require('./lib/_components/@alife/next-field/lib/index.js');
exports.Form = require('./lib/_components/@alife/next-form/lib/index.js');
exports.Grid = require('./lib/_components/@alife/next-grid/lib/index.js');
exports.Icon = require('./lib/_components/@alife/next-icon/lib/index.js');
exports.Input = require('./lib/_components/@alife/next-input/lib/index.js');
exports.Loading = require('./lib/_components/@alife/next-loading/lib/index.js');
exports.LocaleProvider = require('./lib/_components/@alife/next-locale-provider/lib/index.js');
exports.Menu = require('./lib/_components/@alife/next-menu/lib/index.js');
exports.Moment = require('./lib/_components/@alife/next-moment/lib/index.js');
exports.moment = require('./lib/_components/@alife/next-moment/lib/index.js');
exports.Nav = require('./lib/_components/@alife/next-nav/lib/index.js');
exports.Navigation = require('./lib/_components/@alife/next-navigation/lib/index.js');
exports.Notice = require('./lib/_components/@alife/next-notice/lib/index.js');
exports.NumberPicker = require('./lib/_components/@alife/next-number-picker/lib/index.js');
exports.Overlay = require('./lib/_components/@alife/next-overlay/lib/index.js');
exports.Pagination = require('./lib/_components/@alife/next-pagination/lib/index.js');
exports.Progress = require('./lib/_components/@alife/next-progress/lib/index.js');
exports.Radio = require('./lib/_components/@alife/next-radio/lib/index.js');
exports.Range = require('./lib/_components/@alife/next-range/lib/index.js');
exports.Rating = require('./lib/_components/@alife/next-rating/lib/index.js');
exports.Search = require('./lib/_components/@alife/next-search/lib/index.js');
exports.Select = require('./lib/_components/@alife/next-select/lib/index.js');
exports.Slider = require('./lib/_components/@alife/next-slider/lib/index.js');
exports.Step = require('./lib/_components/@alife/next-step/lib/index.js');
exports.Switch = require('./lib/_components/@alife/next-switch/lib/index.js');
exports.Tab = require('./lib/_components/@alife/next-tab/lib/index.js');
exports.Table = require('./lib/_components/@alife/next-table/lib/index.js');
exports.Tag = require('./lib/_components/@alife/next-tag/lib/index.js');
exports.TimePicker = require('./lib/_components/@alife/next-time-picker/lib/index.js');
exports.Timeline = require('./lib/_components/@alife/next-timeline/lib/index.js');
exports.Transfer = require('./lib/_components/@alife/next-transfer/lib/index.js');
exports.Tree = require('./lib/_components/@alife/next-tree/lib/index.js');
exports.TreeSelect = require('./lib/_components/@alife/next-tree-select/lib/index.js');
exports.Upload = require('./lib/_components/@alife/next-upload/lib/index.js');
exports.version = '0.19.22';
